lib/queryBuilder/graphUpserter/UpsertNode.js

```diff
@@ -31,6 +31,16 @@ class UpsertNode {
     const { types, omitFromUpdate } = getTypes(this);
     this.types = types;
 
+    if (parentNode) {
+      const { name } = relExpr;
+      const { relations } = parentNode;
+      const nodes = (relations[name] = relations[name] || []);
+      this.index = nodes.push(this) - 1;
+    } else {
+      this.index = null;
+    }
+    this.dataPath = getDataPath(this);
+
```

Comment:
@koskimas I don't understand why, but this part needed to be moved below the call of `getTypes()` for the tests not to break... Took me a while to figure it out.

Reply:
Actually I think this may be wrong. There can be multiple `UpsertNode`s for one relation. The index may not be correct.

Reply:
Oh damn... Is there a way to fix that in a way that the index points to the right place in the graph always?